Base of RMI with User.md

alopesmendes opened the pull request:
Almost finish the base of rmi still need some changes.
Started writing the user readme on explain the usage.

alopesmendes committed:
modifié :         ifs-cars-server/src/main/java/fr/uge/corp/ifscars/cars/Car.java
	modifié :         ifs-cars-server/src/main/java/fr/uge/corp/ifscars/cars/ICar.java
	modifié :         ifs-cars-server/src/main/java/fr/uge/corp/ifscars/cars/Storage.java
	modifié :         ifs-cars-server/src/main/java/fr/uge/corp/ifscars/rating/Rating.java
	modifié :         ifs-cars-server/src/main/java/fr/uge/corp/ifscars/renting/IRentingService.java
	modifié :         ifs-cars-server/src/main/java/fr/uge/corp/ifscars/renting/RentingService.java
	modifié :         ifs-cars-server/src/main/java/fr/uge/corp/ifscars/server/Main.java
	modifié :         ifs-cars-web/WebContent/WEB-INF/server-config.wsdd
	modifié :         renting-app/src/main/java/fr/uge/corp/renting/Main.java

alopesmendes committed:
rmi with observe

alopesmendes committed:
modifié :         src/main/java/fr/uge/corp/ifscars/cars/Car.java
	modifié :         src/main/java/fr/uge/corp/ifscars/cars/ICar.java
	modifié :         src/main/java/fr/uge/corp/ifscars/cars/Storage.java
	modifié :         src/main/java/fr/uge/corp/ifscars/rating/Rating.java
	modifié :         src/main/java/fr/uge/corp/ifscars/renting/IRentingService.java
	modifié :         src/main/java/fr/uge/corp/ifscars/renting/RentingService.java
	modifié :         ../renting-app/src/main/java/fr/uge/corp/ifscars/cars/ICar.java
	modifié :         ../renting-app/src/main/java/fr/uge/corp/ifscars/cars/Storage.java
	modifié :         ../renting-app/src/main/java/fr/uge/corp/ifscars/rating/Rating.java
	modifié :         ../renting-app/src/main/java/fr/uge/corp/ifscars/renting/IRentingService.java
	modifié :         ../renting-app/src/main/java/fr/uge/corp/renting/Main.java
	modifié :         ../renting-app/src/main/java/fr/uge/corp/rentingapp/employee/Employee.java
	modifié :         ../renting-app/src/main/java/fr/uge/corp/rentingapp/employee/IEmployee.java

alopesmendes committed:
with user readme

Alex1304 commented on ifs-cars-server/src/main/java/fr/uge/corp/ifscars/cars/ICar.java:
```diff
@@ -11,8 +15,117 @@
+			} catch (RemoteException e) {
+				e.printStackTrace();
+			}
+			return null;
```
Fais plutôt `throw new RuntimeException(e);` dans le catch au lieu de print et retourner `null`

Alex1304 commented on ifs-cars-server/src/main/java/fr/uge/corp/ifscars/cars/Car.java:
```diff
@@ -1,29 +1,78 @@
+ * @version 1.8
+ *
+ */
 @SuppressWarnings("serial")
```
je pense c'est mieux d'enlever les `@SuppressWarnings("serial")` et de configurer Eclipse pour ignorer les warnings à la place. Quand on compile à la main (avec javac ou mvn compile) ces warnings n'existent pas.

Alex1304 commented on ifs-cars-server/src/main/java/fr/uge/corp/ifscars/cars/ICar.java:
```diff
@@ -11,8 +15,117 @@
+	 *
+	 */
+	@SuppressWarnings("serial")
+	static class NullCar extends UnicastRemoteObject implements ICar {
```
Hmm, vu l'utilisation que tu fais du `NullCar` je ne vois pas trop ce qui justifie d'avoir ça. Il n'y avait pas moyen d'utiliser `java.util.Optional` à la place? Ça aurait été plus propre.

Alex1304 commented on ifs-cars-server/src/main/java/fr/uge/corp/ifscars/renting/RentingService.java:
```diff
@@ -2,20 +2,158 @@
+			}
+		}
+		return storage.take(model);
+	}
```
Alors bon peut être qu'à première vue ça marche, mais sauf que là on contourne le problème et au final on n'utilise même pas le pattern Observer... Bloquer le serveur avec des queues bloquantes est une très mauvaise idée (et une Map de Queue ça consomme de la mémoire...).

Si on réfléchit à la bonne manière de faire, il faut se rappeler du schéma:
client  ==> serveur : envoi d'une notification pour demander la location d'un modèle
client <== serveur : envoi d'une notification pour donner le véhicule au client une fois dispo
client ==> serveur : envoi d'une notification pour rendre le véhicule, avec le rating et la condition

On voit bien que les échanges de notifications se font dans les deux sens. On pourrait représenter un échange par une méthode dans la classe de celui qui reçoit la notification:
```java
// Lors de la location
server.receiveCarRentingRequest(client, model);
// le code de cette méthode appelle client.receiveCar(car) si disponible
// immédiatement, sinon ajoute le client en liste d'attente et sera traîté plus
// tard par receiveCarReturnRequest.

// Lors de la restitution
server.receiveCarReturnRequest(client, car, rating);
// le code de cette méthode appelle client.receiveCar(car) si un client est
// en attente pour ce modèle, sinon ne fait rien mis à part sauvegarder le rating
```
L'idéal serait de mettre l'appel à `client.receiveCar` dans une méthode privée `processQueue` qui est appelée à la fois dans `server.receiveCarRentingRequest` et dans `server.receiveCarReturnRequest`.

Avec ça, pas besoin de créer bêtement des interfaces IObserve et IObservateur pour au final ne pas les utiliser correctement.

alopesmendes replied:
J'ai implemente les differentes methodes. Les dfferents à noter sont que pour la queue j'ai decide de plutot sauvergarder les temps des demande avec System.currentTime. J'ai vire les classes donc il a seulement les interfaces. ICar il extends toujours Remote puisque j'avais besoin du cote client pour retourne la voiture. J'ai update le readme aussi

Alex1304 commented on renting-app/src/main/java/fr/uge/corp/rentingapp/employee/IEmployee.java:
```diff
@@ -0,0 +1,57 @@
+ * @version 1.8
+ *
+ */
+public interface IEmployee {
```
T'as oublié le `extends Remote`

Alex1304 reviewed (changes requested):
Il y a certains choix de design qui peuvent vraiment être revus, sinon c'est un bon début.

En particulier:
* Revoir de quel côté (client ou serveur) doivent se situer les implémentations des interfaces RMI. Je vois que tu as dupliqué les implémentations de Car, de Rating et de Storage dans les deux projets et ça c'est pas bon. Seules les interfaces sont copiées-collées entre le client et le serveur, l'implémentation ne doit être que d'un seul côté (en l’occurrence celui qui fait le `new`). Pour Storage ce sera à priori le serveur, idem pour Car car le client ne fait que les récupérer par modèle. Pour Rating en revanche ce serait côté client, le serveur ne fait que lire les ratings que les clients lui donnent.
* Revoir l'implémentation de RatingService en utilisant correctement le pattern Observer au lieu d'utiliser des blocking queues
* Revoir l'utilité ou non de mettre certaines classes en Remote, par exemple je vois pas pourquoi Rating ou Employee devraient être remotes (je ne sais même pas à quoi sert Employee, on pourrait simplement passer les IDs pour les employés ou faire des passages par valeur pour Rating. Faut voir ce qui justifie ces choix).

Faudra sûrement modifier certaines choses dans l'UML aussi pour pouvoir s'y retrouver.

alopesmendes committed:
modifié :         renting-app/src/main/java/fr/uge/corp/rentingapp/employee/IEmployee.java

alopesmendes committed:
modifié :         ifs-cars-server/src/main/java/fr/uge/corp/ifscars/cars/ICar.java

alopesmendes committed:
modifié :         ifs-cars-server/src/main/java/fr/uge/corp/ifscars/cars/Car.java
	modifié :         ifs-cars-server/src/main/java/fr/uge/corp/ifscars/cars/ICar.java
	modifié :         ifs-cars-server/src/main/java/fr/uge/corp/ifscars/cars/Storage.java
	modifié :         ifs-cars-server/src/main/java/fr/uge/corp/ifscars/observe/Observateur.java
	modifié :         ifs-cars-server/src/main/java/fr/uge/corp/ifscars/rating/Rating.java
	modifié :         ifs-cars-server/src/main/java/fr/uge/corp/ifscars/renting/RentingService.java
	modifié :         ifs-cars-server/src/main/java/fr/uge/corp/ifscars/server/Main.java
	ifs-cars-server/src/main/java/fr/uge/corp/ifscars/cars/IStorage.java

alopesmendes committed:
serial resolve

alopesmendes committed:
modifié :         ifs-cars-server/src/main/java/fr/uge/corp/ifscars/cars/Car.java
	modifié :         ifs-cars-server/src/main/java/fr/uge/corp/ifscars/cars/ICar.java
	supprimé :        ifs-cars-server/src/main/java/fr/uge/corp/ifscars/cars/IStorage.java
	modifié :         ifs-cars-server/src/main/java/fr/uge/corp/ifscars/cars/Storage.java
	supprimé :        ifs-cars-server/src/main/java/fr/uge/corp/ifscars/observe/IObservateur.java
	supprimé :        ifs-cars-server/src/main/java/fr/uge/corp/ifscars/observe/IObserve.java
	supprimé :        ifs-cars-server/src/main/java/fr/uge/corp/ifscars/observe/Observateur.java
	supprimé :        ifs-cars-server/src/main/java/fr/uge/corp/ifscars/rating/Rating.java
	modifié :         ifs-cars-server/src/main/java/fr/uge/corp/ifscars/renting/IRentingService.java
	modifié :         ifs-cars-server/src/main/java/fr/uge/corp/ifscars/renting/RentingService.java
	modifié :         ifs-cars-server/src/main/java/fr/uge/corp/ifscars/server/Main.java
	modifié :         renting-app/src/main/java/fr/uge/corp/ifscars/cars/ICar.java
	supprimé :        renting-app/src/main/java/fr/uge/corp/ifscars/cars/Storage.java
	supprimé :        renting-app/src/main/java/fr/uge/corp/ifscars/observe/IObservateur.java
	supprimé :        renting-app/src/main/java/fr/uge/corp/ifscars/observe/IObserve.java
	modifié :         renting-app/src/main/java/fr/uge/corp/ifscars/rating/Rating.java
	modifié :         renting-app/src/main/java/fr/uge/corp/ifscars/renting/IRentingService.java
	modifié :         renting-app/src/main/java/fr/uge/corp/renting/Main.java
	supprimé :        renting-app/src/main/java/fr/uge/corp/rentingapp/employee/Employee.java
	supprimé :        renting-app/src/main/java/fr/uge/corp/rentingapp/employee/IEmployee.java

alopesmendes committed:
modifié :         Images/UML.png
	modifié :         ifs-cars-server/src/main/java/fr/uge/corp/ifscars/cars/Storage.java
	nouveau fichier : ifs-cars-server/src/main/java/fr/uge/corp/ifscars/rating/IRating.java
	modifié :         ifs-cars-server/src/main/java/fr/uge/corp/ifscars/renting/IRentingService.java
	modifié :         ifs-cars-server/src/main/java/fr/uge/corp/ifscars/renting/RentingService.java
	nouveau fichier : ifs-cars-server/src/main/java/fr/uge/corp/rentingapp/client/IClient.java
	nouveau fichier : renting-app/src/main/java/fr/uge/corp/ifscars/rating/IRating.java
	nouveau fichier : renting-app/src/main/java/fr/uge/corp/rentingapp/client/Client.java
	nouveau fichier : renting-app/src/main/java/fr/uge/corp/rentingapp/client/IClient.java

alopesmendes committed:
modifié :         Images/UML.png

alopesmendes committed:
modifié :         ifs-cars-server/src/main/java/fr/uge/corp/ifscars/cars/Storage.java
	modifié :         ifs-cars-server/src/main/java/fr/uge/corp/ifscars/rating/IRating.java
	modifié :         ifs-cars-server/src/main/java/fr/uge/corp/ifscars/renting/IRentingService.java
	modifié :         ifs-cars-server/src/main/java/fr/uge/corp/ifscars/renting/RentingService.java
	modifié :         ifs-cars-server/src/main/java/fr/uge/corp/rentingapp/client/IClient.java
	modifié :         renting-app/src/main/java/fr/uge/corp/rentingapp/client/IClient.java

Alex1304 commented:
Ok merci on va voir plus tard ce qui peut être encore amélioré, pour l'instant c'est ok
